br/pkg/restore/import.go

```diff
@@ -380,39 +384,92 @@ func (importer *FileImporter) ClearFiles(ctx context.Context, pdClient pd.Client
+	if r != nil && r.Region != nil {
+		for i, f := range files {
+			if bytes.Compare(r.Region.StartKey, ranges[i].EndKey) <= 0 &&
+				(len(r.Region.EndKey) == 0 || bytes.Compare(r.Region.EndKey, ranges[i].StartKey) >= 0) {
+				output = append(output, f)
+			}
+		}
+	} else {
+		output = files
+	}
+
+	return output, nil
+}
+
 // ImportKVFiles restores the kv events.
 func (importer *FileImporter) ImportKVFiles(
 	ctx context.Context,
-	file *backuppb.DataFileInfo,
+	files []*backuppb.DataFileInfo,
 	rule *RewriteRules,
+	shiftStartTS uint64,
 	startTS uint64,
 	restoreTS uint64,
+	supportBatch bool,
 ) error {
-	startTime := time.Now()
-	log.Debug("import kv files", zap.String("file", file.Path))
-	startKey, endKey, err := GetRewriteEncodedKeys(file, rule)
-	if err != nil {
-		return errors.Trace(err)
+	var (
+		startTime = time.Now()
+		startKey  []byte
+		endKey    []byte
+		ranges    = make([]kv.KeyRange, len(files))
+		err       error
+	)
+
+	if !supportBatch && len(files) > 1 {
+		return errors.Annotatef(berrors.ErrInvalidArgument,
+			"do not support batch apply but files count:%v > 1", len(files))
+	}
+	log.Debug("import kv files", zap.Int("batch file count", len(files)))
+
+	for i, f := range files {
+		ranges[i].StartKey, ranges[i].EndKey, err = GetRewriteEncodedKeys(f, rule)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if len(startKey) == 0 || bytes.Compare(ranges[i].StartKey, startKey) < 0 {
+			startKey = ranges[i].StartKey
+		}
+		if len(endKey) == 0 || bytes.Compare(ranges[i].EndKey, endKey) > 0 {
+			endKey = ranges[i].EndKey
+		}
 	}
 
 	log.Debug("rewrite file keys",
-		zap.String("name", file.Path),
-		logutil.Key("startKey", startKey),
-		logutil.Key("endKey", endKey))
+		logutil.Key("startKey", startKey), logutil.Key("endKey", endKey))
 
-	// This RetryState will retry 48 time, for 5 min - 6 min.
-	rs := utils.InitialRetryState(48, 100*time.Millisecond, 8*time.Second)
+	// This RetryState will retry 45 time, about 10 min.
+	rs := utils.InitialRetryState(45, 100*time.Millisecond, 15*time.Second)
 	ctl := OverRegionsInRange(startKey, endKey, importer.metaClient, &rs)
 	err = ctl.Run(ctx, func(ctx context.Context, r *split.RegionInfo) RPCResult {
-		return importer.ImportKVFileForRegion(ctx, file, rule, startTS, restoreTS, r)
+		subfiles, errFilter := FilterFilesByRegion(files, ranges, r)
+		if errFilter != nil {
+			return RPCResultFromError(errFilter)
+		}
+		if len(subfiles) == 0 {
+			return RPCResultOK()
+		}
+		return importer.ImportKVFileForRegion(ctx, subfiles, rule, shiftStartTS, startTS, restoreTS, r, supportBatch)
 	})
 
-	log.Debug("download and apply file done",
-		zap.String("file", file.Path),
-		zap.Stringer("take", time.Since(startTime)),
-		logutil.Key("fileStart", file.StartKey),
-		logutil.Key("fileEnd", file.EndKey),
-	)
+	for _, file := range files {
+		log.Debug("download and apply file done",
+			zap.String("file", file.Path), zap.Stringer("take", time.Since(startTime)),
```

Comment:
it seems enough to call `time.Since` once.

Reply:
(BTW, I think we'd better don't execute the whole for block if the log level is less than debug)